facefusion/processors/frame/modules/lip_sync.py

```diff
@@ -0,0 +1,222 @@
+		crop_mask = numpy.minimum.reduce(crop_mask_list)
+		paste_frame = paste_back(temp_frame, crop_frame, crop_mask, affine_matrix)
+	return paste_frame
+
+
+def prepare_audio_frame(audio_frame : AudioFrame) -> AudioFrame:
+	audio_frame = numpy.maximum(numpy.exp(-5 * numpy.log(10)), audio_frame)
+	audio_frame = numpy.log10(audio_frame) * 1.6 + 3.2
+	audio_frame = audio_frame.clip(-4, 4).astype(numpy.float32)
+	audio_frame = numpy.expand_dims(audio_frame, axis = (0, 1))
+	return audio_frame
+
+
+def prepare_crop_frame(crop_frame : VisionFrame) -> VisionFrame:
+	crop_frame = numpy.expand_dims(crop_frame, axis = 0)
+	crop_frame_masked = crop_frame.copy()
+	crop_frame_masked[:, 48:] = 0
+	crop_frame_stack = numpy.concatenate((crop_frame_masked, crop_frame), axis = 3)
+	crop_frame_stack = crop_frame_stack.transpose(0, 3, 1, 2).astype('float32') / 255.0
+	return crop_frame_stack
+
+
+def normalize_crop_frame(crop_frame : VisionFrame) -> VisionFrame:
+	crop_frame = crop_frame[0].transpose(1, 2, 0)
+	crop_frame = crop_frame.clip(0, 1) * 255
+	crop_frame = crop_frame.astype(numpy.uint8)
+	return crop_frame
+
+
+def get_reference_frame(source_face : Face, target_face : Face, temp_frame : VisionFrame) -> VisionFrame:
+	pass
+
+
+def process_frame(audio_frame : numpy.ndarray[Any, Any], reference_faces : FaceSet, temp_frame : VisionFrame) -> VisionFrame:
+	if 'reference' in facefusion.globals.face_selector_mode:
+		similar_faces = find_similar_faces(temp_frame, reference_faces, facefusion.globals.reference_face_distance)
+		if similar_faces:
+			for similar_face in similar_faces:
+				temp_frame = lip_sync(audio_frame, similar_face, temp_frame)
+	if 'one' in facefusion.globals.face_selector_mode:
+		target_face = get_one_face(temp_frame)
+		if target_face:
+			temp_frame = lip_sync(audio_frame, target_face, temp_frame)
+	if 'many' in facefusion.globals.face_selector_mode:
+		many_faces = get_many_faces(temp_frame)
+		if many_faces:
+			for target_face in many_faces:
+				temp_frame = lip_sync(audio_frame, target_face, temp_frame)
+	return temp_frame
+
+
+def process_frames(source_paths : List[str], temp_frame_paths : List[str], update_progress : Update_Process) -> None:
+	reference_faces = get_reference_faces() if 'reference' in facefusion.globals.face_selector_mode else None
+	source_audio_path = filter_audio_paths(source_paths)[0]
+	video_fps = detect_video_fps(facefusion.globals.target_path)
+	for temp_frame_path in temp_frame_paths:
+		frame_number = int(os.path.basename(temp_frame_path).split(".")[0])
+		audio_frame = get_audio_frame(source_audio_path, video_fps, frame_number)
+		if audio_frame is not None:
+			temp_frame = read_image(temp_frame_path)
+			result_frame = process_frame(audio_frame, reference_faces, temp_frame)
+			write_image(temp_frame_path, result_frame)
+		update_progress()
+
+
+def process_image(source_paths : List[str], target_path : str, output_path : str) -> None:
+	# TODO
```

Comment:
the output expects an images... so I suggest we integrate kinda stupid solution.

one vision frame + one audio frame => image